### PR TITLE
Feat: CatmullClark subdivision surface for sphere in primitive mesh

### PR DESCRIPTION
**Pull request by JujieX**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
CatmullClark subdivision surface for sphere generation in primitive mesh
fix https://github.com/galacean/editor/issues/1628

### CODE REVIEW

**Comment by zhuxudong** [comment]
1. uv has some problems
image
2. when the new algorithm takes 4 and the old algorithm takes 64, it can almost get a smooth spherical shape, but the number of faces is more
`createSphere(engine ,1, 64)` vs `createSphere(engine, 1, 4)` 
image
image

3. may skyProcedural's bug, when the step is set to 5, the skyline is almost clear, but the number of faces is more than before, and there is a blue anomaly
image

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 1]
```diff
@@ -336,6 +430,271 @@ export class PrimitiveMesh {
...
+        const cell = cells[i];
+
+        faces[i] = {
+          facePoint: new Vector3(),
+          vertices: [],
+          edges: []
+        };
+
+        // get cell points
+        for (let j = 0; j < cell.length; j++) {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:486]: cache `n`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 2]
```diff
@@ -336,6 +430,271 @@ export class PrimitiveMesh {
...
+      for (const key in edges) {
+        const edge = edges[key];
+
+        for (let j = 0; j < edge.faces.length; j++) {
+          const curFace = faces[edge.faces[j]];
+          edge.edgePoint.add(curFace.facePoint);
+        }
+        edge.edgePoint.scale(0.5);
+        edge.edgePoint.add(edge.midPoint);
+        edge.edgePoint.scale(0.5);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:545]: Cache `edgePoint` and `faces`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 3]
```diff
@@ -336,6 +430,271 @@ export class PrimitiveMesh {
...
+
+        // get cell face's facePoint
+        faces[i].facePoint.scale(1 / cell.length);
+
+        // get cell edges
+        for (let j = 0; j < cell.length; j++) {
+          const vertexIdxA = cell[j];
+          const vertexIdxB = cell[(j + 1) % cell.length];
+          const edgeIdxKey = [vertexIdxA, vertexIdxB].sort().toString();
+
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:513]: Use number compare, performance waste, ignore use string as key, use number directly

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 4]
```diff
@@ -336,6 +430,271 @@ export class PrimitiveMesh {
...
+    const getIndex = (vertex: Vector3): number => {
+      const idx = vertices.findIndex((value) => Vector3.equals(value, vertex));
+      return idx > -1 ? idx : vertices.push(vertex) - 1;
+    };
+
+    for (let i = 0; i < step; i++) {
+      let points: Array<IPoint> = [];
+      let edges: Array<IEdge> = [];
+      let faces: Array<IFace> = [];
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:473]: Try to reuse it in loop

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 5]
```diff
@@ -257,6 +286,71 @@ export class PrimitiveMesh {
...
+        (Math.PI - Math.atan2(PrimitiveMesh._tempVec30.z, PrimitiveMesh._tempVec30.x)) / (2 * Math.PI);
+      vertices[i * 8 + 7] = Math.acos(PrimitiveMesh._tempVec30.y / PrimitiveMesh._tempVec30.length()) / Math.PI;
+    }
+
+    // get indices
+    let offset = 0;
+    for (let i = 0; i < cells.length; i++) {
+      const cell = cells[i];
+
+      // 这个指的是positions之前的
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:331]: Use english

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 6]
```diff
@@ -257,6 +286,71 @@ export class PrimitiveMesh {
...
+    const vertexCount = positions.length + Math.pow(2, step + 1) + 15;
+    const vertices = new Float32Array(vertexCount * 8);
+    const indices = PrimitiveMesh._generateIndices(sphereMesh.engine, positions.length, cells.length * 6);
+
+    // get normals, uvs, and scale to radius
+    for (let i = 0; i < positions.length; i++) {
+      PrimitiveMesh._tempVec30.copyFrom(positions[i]);
+      PrimitiveMesh._tempVec30.normalize();
+
+      vertices[i * 8] = PrimitiveMesh._tempVec30.x * radius;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:313]: cache `_tempVec30` before loop

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 7]
```diff
@@ -21,6 +22,30 @@ import { VertexAttribute } from "./enums/VertexAttribute";
  */
 export class PrimitiveMesh {
   private static _tempVec30: Vector3 = new Vector3();
+  private static _tempVec31: Vector3 = new Vector3();
+  private static _tempVec32: Vector3 = new Vector3();
+  private static _tempVec33: Vector3 = new Vector3();
+  private static _tempVec34: Vector3 = new Vector3();
+
+  private static readonly _initialVertices = [
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:30]: Unclear naming

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 8]
```diff
@@ -21,6 +22,30 @@ import { VertexAttribute } from "./enums/VertexAttribute";
...
+    new Vector3(-1, -1, 1),
+    new Vector3(1, -1, 1),
+    new Vector3(1, 1, 1),
+    new Vector3(1, -1, -1),
+    new Vector3(1, 1, -1),
+    new Vector3(-1, -1, -1),
+    new Vector3(-1, 1, -1)
+  ];
+
+  private static readonly _initialCells = [
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:41]: Unclear naming

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 9]
```diff
@@ -46,6 +71,30 @@ export class PrimitiveMesh {
...
 
+  /**
+   * Create a sphere mesh by implementing Catmull-Clark Surface Subdivision Algorithm.
+   * @param engine - Engine
+   * @param radius - Sphere radius
+   * @param step - Number of subdiv steps
+   * @param noLongerAccessible - No longer access the vertices of the mesh after creation
+   * @returns Sphere model mesh
+   */
+  static createCCSubdivSphere(
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:82]: `createSubdivisionSurfaceSphere` is better

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 10]
```diff
@@ -336,6 +452,262 @@ export class PrimitiveMesh {
     PrimitiveMesh._initialize(sphereMesh, vertices, indices, noLongerAccessible, isRestoreMode, restoreVertexBuffer);
   }
 
+  /**
+   * @internal
+   */
+  static _subdivCatmullClark(step: number): {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:458]: void?

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 11]
```diff
@@ -336,6 +452,262 @@ export class PrimitiveMesh {
...
+    cells: number[][];
+    positions: Vector3[];
+  } {
+    const { _tempVec31: tempVec1, _tempVec32: tempVec2, _tempVec33: tempVec3 } = PrimitiveMesh;
+
+    let cells = PrimitiveMesh._initialCells.map((cell) => [...cell]);
+    let vertices = PrimitiveMesh._initialVertices.map((vertex) => vertex.clone());
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:465]:
```ts 
 // Create Seed cells and vertices data
    const cells = [
      [0, 1, 2, 3],
      [3, 2, 4, 5],
      [5, 4, 6, 7],
      [7, 0, 3, 5],
      [7, 6, 1, 0],
      [6, 4, 2, 1]
    ];
    const vertices = [
      new Vector3(-1, 1, 1),
      new Vector3(-1, -1, 1),
      new Vector3(1, -1, 1),
      new Vector3(1, 1, 1),
      new Vector3(1, -1, -1),
      new Vector3(1, 1, -1),
      new Vector3(-1, -1, -1),
      new Vector3(-1, 1, -1)
    ];
```

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 10 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts:456]: :{  cells: Float32Array;
    positions: Float32Array;
}

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 12]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+    sphereMesh: ModelMesh,
+    radius: number,
+    step: number,
+    noLongerAccessible: boolean,
+    isRestoreMode: boolean,
+    restoreVertexBuffer?: Buffer
+  ): void {
+    step = MathUtil.clamp(Math.floor(step), 1, 6);
+
+    let { positions, cells } = PrimitiveMesh._subdivCatmullClark(step);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:319]: `const { positions, cells }`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 13]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+    const positionCount = positions.length / 3;
+    const cellsCount = cells.length * 0.25;
+    const vertexCount = positionCount + Math.pow(2, step + 1) + 15;
+
+    const vertices = new Float32Array(vertexCount * 8);
+    const indices = PrimitiveMesh._generateIndices(sphereMesh.engine, positionCount, cellsCount * 6);
+
+    // Get normals, uvs, and scale to radius.
+    for (let i = 0; i < positionCount; i++) {
+      let x = positions[3 * i];
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:330]: cache `3 * i`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 14]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+      let y = positions[3 * i + 1];
+      let z = positions[3 * i + 2];
+
+      let len = Math.sqrt(x * x + y * y + z * z);
+      len = 1 / len;
+      x = x * len;
+      y = y * len;
+      z = z * len;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:338]:
```ts
x *= len;
y *= len;
z *= len;
```

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 15]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+      vertices[i * 8 + 1] = y * radius;
+      vertices[i * 8 + 2] = z * radius;
+
+      vertices[i * 8 + 3] = x;
+      vertices[i * 8 + 4] = y;
+      vertices[i * 8 + 5] = z;
+
+      vertices[i * 8 + 6] = (Math.PI - Math.atan2(z, x)) / (2 * Math.PI);
+      vertices[i * 8 + 7] = Math.acos(y) / Math.PI;
+    }
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:350]:
```suggestion
    for (let i = 0; i < positionCount; i++) {
      let offset = 3 * i;
      let x = positions[offset];
      let y = positions[offset + 1];
      let z = positions[offset + 2];

      const reciprocalLength = 1 / Math.sqrt(x * x + y * y + z * z);
      x *= reciprocalLength;
      y *= reciprocalLength;
      z *= reciprocalLength;

      offset = i * 8;
      vertices[offset] = x * radius;
      vertices[offset + 1] = y * radius;
      vertices[offset + 2] = z * radius;

      vertices[offset + 3] = x;
      vertices[offset + 4] = y;
      vertices[offset + 5] = z;

      vertices[offset + 6] = (Math.PI - Math.atan2(z, x)) / (2 * Math.PI);
      vertices[offset + 7] = Math.acos(y) / Math.PI;
    }
```

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 16]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+
+    // Get indices.
+    let offset = 0;
+    for (let i = 0; i < cellsCount; i++) {
+      indices[offset++] = cells[4 * i];
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:355]:
```ts
// Get indices
    for (let i = 0; i < cellsCount; i++) {
      let offset = 0;
```

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 17]
```diff
@@ -336,6 +453,338 @@ export class PrimitiveMesh {
...
+        let bIdx0 = 0,
+          bIdx1 = 0,
+          bIdx2 = 0,
+          dIdx0 = 0;
+
+        for (let j = 0; j < 4; j++) {
+          // Get the updated exising point index.
+          const ia = preCells[pointIdx++];
+
+          let id, ib;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:610]: no type

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 18]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+    isRestoreMode: boolean,
+    restoreVertexBuffer?: Buffer
+  ): void {
+    step = MathUtil.clamp(Math.floor(step), 1, 6);
+
+    let { positions, cells } = PrimitiveMesh._subdivCatmullClark(step);
+
+    const positionCount = positions.length / 3;
+    const cellsCount = cells.length * 0.25;
+    const vertexCount = positionCount + Math.pow(2, step + 1) + 15;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:323]: 15

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 19]
```diff
@@ -257,6 +303,77 @@ export class PrimitiveMesh {
...
+    noLongerAccessible: boolean,
+    isRestoreMode: boolean,
+    restoreVertexBuffer?: Buffer
+  ): void {
+    step = MathUtil.clamp(Math.floor(step), 1, 6);
+
+    let { positions, cells } = PrimitiveMesh._subdivCatmullClark(step);
+
+    const positionCount = positions.length / 3;
+    const cellsCount = cells.length * 0.25;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:322]: cells.length / 0.25;

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 20]
```diff
@@ -336,6 +456,352 @@ export class PrimitiveMesh {
...
+    cells: Float32Array;
+    positions: Float32Array;
+  } {
+    const { _tempVec31: tempVec1, _tempVec32: tempVec2, _tempVec33: tempVec3 } = PrimitiveMesh;
+
+    const points: Array<IPoint> = [];
+    const edges: Map<number, IEdge> = new Map();
+    const faces: Array<IFace> = [];
+
+    let previousPositions = PrimitiveMesh._sphereSeedPositions.slice();
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:472]:
```ts
for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
        for (int k = 0; k < o; k++) {
            // 代码块
        }
    }
}
```

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 21]
```diff
@@ -336,6 +456,352 @@ export class PrimitiveMesh {
     PrimitiveMesh._initialize(sphereMesh, vertices, indices, noLongerAccessible, isRestoreMode, restoreVertexBuffer);
   }
 
+  /**
+   * @internal
+   */
+  static _subdivCatmullClark(step: number): {
+    cells: Float32Array;
+    positions: Float32Array;
+  } {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:465]: you can try multi out params to ignore new Object.
```ts
static _subdivCatmullClark(step: number, out: (positions: Float32Array, cells: Float32Array) => void): void {
```

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 22]
```diff
@@ -336,6 +456,352 @@ export class PrimitiveMesh {
...
+    const { _tempVec31: tempVec1, _tempVec32: tempVec2, _tempVec33: tempVec3 } = PrimitiveMesh;
+
+    const points: Array<IPoint> = [];
+    const edges: Map<number, IEdge> = new Map();
+    const faces: Array<IFace> = [];
+
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:471]: Uniform use of type inference：

```suggestion
     const points = new Array<IPoint>();
    const edges = new Map<number, IEdge>();
    const faces = new Array<IFace>();
```

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 12 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 13 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 14 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 15 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts:352]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 20 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts:472]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 19 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 17 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 23]
```diff
@@ -336,6 +456,353 @@ export class PrimitiveMesh {
...
+      edges.clear();
+      faces.length = 0;
+
+      // Get cell faces.
+      for (let j = 0; j < preCellCount; j++) {
+        faces[j] = {
+          facePoint: new Vector3(),
+          vertices: [],
+          adjacentEdges: []
+        };
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:493]: cache face

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 22 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts:471]: done

File: packages/core/src/mesh/PrimitiveMesh.ts (hunk 23 above)
Comment (JujieX) [reply packages/core/src/mesh/PrimitiveMesh.ts]: done

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 24]
```diff
@@ -336,6 +456,353 @@ export class PrimitiveMesh {
...
+
+        for (let k = 0; k < 4; k++) {
+          const idx = preCells[4 * j + k];
+          const offset = 3 * idx;
+          const vertex = new Vector3(positions[offset], positions[offset + 1], positions[offset + 2]);
+
+          if (!points[idx]) {
+            const point: IPoint = {
+              position: vertex,
+              newPosition: new Vector3(),
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:503]: Never used

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 25]
```diff
@@ -336,6 +456,306 @@ export class PrimitiveMesh {
...
+        });
+
+        // Get cell face's facePoint.
+        for (let k = 0; k < 4; k++) {
+          const offset = 3 * preCells[4 * j + k];
+          face.facePoint.x += positions[offset];
+          face.facePoint.y += positions[offset + 1];
+          face.facePoint.z += positions[offset + 2];
+        }
+        face.facePoint.scale(0.25);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:497]: Why not merge facePoint and edge compute code

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 26]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+  } {
+    const edges = new Map<number, IEdge>();
+    const faces = new Array<IFace>();
+
+    let previousPositions = PrimitiveMesh._sphereSeedPositions.slice();
+    let preCells = PrimitiveMesh._sphereSeedCells.slice();
+
+    for (let i = 0; i < step; i++) {
+      const positionCount = 24 * Math.pow(4, i) + 2;
+      const positions = new Float32Array(3 * positionCount);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:468]: We can create a bug buffer directly!

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 27]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+      positions.set(previousPositions);
+
+      const preCellCount = preCells.length * 0.25;
+      const cells = new Float32Array(24 * Math.pow(4, i + 1));
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:472]: `positions` and `cells` is better

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 28]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+
+          edgePoint.x += 0.25 * facePoint.x;
+          edgePoint.y += 0.25 * facePoint.y;
+          edgePoint.z += 0.25 * facePoint.z;
+        }
+      }
+
+      const prePointCount = 6 * Math.pow(4, i) + 2;
+      const facePointCount = 6 * Math.pow(4, i);
+      const edgePointoffset = prePointCount + facePointCount;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:531]: `edgePointOffset`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 29]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+          const edgePoint = edge.edgePoint;
+          const facePoint = face.facePoint;
+
+          edgePoint.x += 0.25 * facePoint.x;
+          edgePoint.y += 0.25 * facePoint.y;
+          edgePoint.z += 0.25 * facePoint.z;
+        }
+      }
+
+      const prePointCount = 6 * Math.pow(4, i) + 2;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:529]: cache `6 * Math.pow(4, i)`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 30]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+              temp = id;
+              break;
+            }
+            case 1: {
+              const edgeB = face.adjacentEdges[k % 4];
+              id = ib;
+              ib = this._calculateEdgeIndex(positions, edgeB, edgePointoffset);
+              break;
+            }
+            case 2: {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:572]: case 1 and case 2 is same

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 31]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+   * @internal
+   */
+  static _subdivCatmullClark(step: number): {
+    cells: Float32Array;
+    positions: Float32Array;
+  } {
+    const edges = new Map<number, IEdge>();
+    const faces = new Array<IFace>();
+
+    let previousPositions = PrimitiveMesh._sphereSeedPositions.slice();
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:463]: `edges` and `faces` can use big pool

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 32]
```diff
@@ -257,6 +296,81 @@ export class PrimitiveMesh {
...
+    radius: number,
+    step: number,
+    noLongerAccessible: boolean,
+    isRestoreMode: boolean,
+    restoreVertexBuffer?: Buffer
+  ): void {
+    // Max step is limited to 6. Because 7 step will generate a single mesh with over 98306 vertices
+    step = MathUtil.clamp(Math.floor(step), 1, 6);
+
+    const { positions, cells } = PrimitiveMesh._subdivCatmullClark(step);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:313]: `positions` and `cells` can create outside once

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 33]
```diff
@@ -257,6 +296,81 @@ export class PrimitiveMesh {
...
+
+    const positionCount = positions.length / 3;
+    const cellsCount = cells.length / 4;
+    const vertexCount = positionCount + Math.pow(2, step + 1) + 15;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:317]: 15 can opt

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 34]
```diff
@@ -257,6 +296,81 @@ export class PrimitiveMesh {
...
+      indices[offset++] = cells[idx + 2];
+
+      indices[offset++] = cells[idx];
+      indices[offset++] = cells[idx + 2];
+      indices[offset++] = cells[idx + 3];
+    }
+
+    // Solve texture seam problem caused by vertex sharing
+    const count = PrimitiveMesh._computeFlippedVertex(indices, vertices, positionCount);
+    PrimitiveMesh._computePolesVertex(indices, vertices, positionCount + count);
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:363]: opt `_computeFlippedVertex` and `_computePolesVertex`  naming

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 35]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+      }
+    }
+    return count;
+  }
+
+  /**
+   * @internal
+   * Duplicate vertices at the poles and adjust their UV coordinates.
+   */
+  static _computePolesVertex(indices: Uint16Array | Uint32Array, vertices: Float32Array, currentCount: number): void {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:658]: private

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 36]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+      preCells = cells;
+    }
+    return { cells: preCells, positions: previousPositions };
+  }
+
+  /**
+   * @internal
+   * Duplicate vertices whose uv normal is flipped and adjust their UV coordinates.
+   */
+  static _computeFlippedVertex(
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:603]: private

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 37]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+    indices: Uint16Array | Uint32Array,
+    vertices: Float32Array,
+    currentCount: number
+  ): number {
+    const flippedVertex: Set<number> = new Set();
+
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:609]: `const flippedVertex = new Set<number>();`

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 38]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+              vertices[offset] = vertices[idx];
+              vertices[offset + 1] = vertices[idx + 1];
+              vertices[offset + 2] = vertices[idx + 2];
+
+              vertices[offset + 3] = vertices[idx + 3];
+              vertices[offset + 4] = vertices[idx + 4];
+              vertices[offset + 5] = vertices[idx + 5];
+
+              vertices[offset + 6] = vertices[idx + 6] + 1;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:640]: why add 1

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 39]
```diff
@@ -336,6 +450,263 @@ export class PrimitiveMesh {
...
+              vertices[offset + 7] = vertices[idx + 7];
+
+              flippedVertex[e] = currentCount + count;
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:643]: why not do this when generate vertex

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 40]
```diff
@@ -46,6 +57,34 @@ export class PrimitiveMesh {
...
+   * Max step is limited to 6.
+   * @param engine - Engine
+   * @param radius - Sphere radius
+   * @param step - Number of subdiv steps
+   * @param noLongerAccessible - No longer access the vertices of the mesh after creation
+   * @returns Sphere model mesh
+   */
+  static createSubdivisionSurfaceSphere(
+    engine: Engine,
+    radius: number = 1,
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:71]: default radius should `0.5``

File: packages/core/src/mesh/PrimitiveMesh.ts [hunk 41]
```diff
@@ -336,6 +450,260 @@ export class PrimitiveMesh {
...
+          count++;
+        }
+      }
+    }
+  }
+  /**
+   * @internal
+   * Get edge point index for subdivision surface sphere.
+   */
+  static _calculateEdgeIndex(positions: Float32Array, edge: IEdge, offset: number): number {
```
Comment (GuoLei1990) [inline packages/core/src/mesh/PrimitiveMesh.ts:695]: private

**Pull request merged**